xtraplatform-tiles/src/main/java/de/ii/xtraplatform/tiles/domain/MbtilesTileset.java

```diff
@@ -386,9 +385,13 @@ public Optional<InputStream> getTile(TileQuery tile) throws SQLException, IOExce
                     : rs.getBinaryStream("tile_data"));
       }
     } catch (SQLException e) {
-      throw new IllegalStateException(String.format("Query execution failed: %s", sql), e);
+      throw new IllegalStateException(
+          String.format(
+              "Could not fetch tile %d/%d/%d. Query execution failed: %s", level, row, col, sql),
+          e);
```

Comment:
@cportele 
After further consideration, I am not convinced the changes will have the desired result. I think this catch clause has to be removed, otherwise the `IllegalStateException` will still stop the thread. When removing the catch clause, the `SQLException` will then be catched in `TileStoreMbTiles.get` and a regular error result is returned. This regular error result will then be logged as warning in `TileCache.doSeed`.

Another question would be why the "database is locked" errors when reading happen in the first place. My assumption would be that reads should not be allowed during writes. If that is the case, we should use a `ReadWriteLock` instead of the `Semaphore`.

Reply:
@azahnen 

With respect to the locking issue: Yes, `locking_mode` is currently set to [`EXCLUSIVE`](https://github.com/interactive-instruments/xtraplatform-spatial/blob/master/xtraplatform-tiles/src/main/java/de/ii/xtraplatform/tiles/domain/MbtilesTileset.java#L90). 

My impression from the tests was that read access still works during seeding (or on-demand caching), that is that SQLite waits sufficiently long on a read request in case of a lock until it is released. But there are probably limits to that. One option could be to try setting the `journal_mode` to [`WAL`](https://www.sqlite.org/wal.html), which should support read access during writes, if I understand it correctly.

Another, potentially complementary option could be retry a read access a few times in case of an error like in the write case.

Reply:
@cportele On the first issue: that might be true, but the IllegalStateException is still the wrong way to communicate that. When removing the catch clause, the only caller TileStoreMbTiles.get will create and return a TileResult.error, which is the proper way to communicate an error to upper layers.

Reply:
@cportele On the second issue: I am just making assumptions since I was not able to reproduce the issue. But since it is still happening for the customer, we should try to further improve the behaviour. And from the error message they provided I am pretty sure that the cause is a read access triggered by the seeding.

Reply:
@azahnen - Please have another look. I have pushed some changes to address the comments including setting the journal mode to `WAL`. There should only be two cases of read accesses to an immutable cache during seeding outside of the mutex: 

- The check, if a tile exists and needs to be reseeded. If that fails, a warning is logged and the tile is reseeded. If that is a very rare case, this should not have a negative performance impact and the cache will still be correct.
- Accessing single-layer tiles in the `combine()`. It now tries to get the tile three times and ignores the layer, if all three attempts fail. A warning is logged. That tile will not be correct, but usable.

I am testing the code, so far without issues (but then again, I also had no issues before those changes).

Reply:
@cportele Thanks, now I am pretty sure that the changes will fix the thread stoppage issue and optimistic that the locking issues might be gone completely. 
I will then proceed to provide a test version to the customer.

Reply:
I have made one more change after a bit of research, I changed `synchronous` from `OFF` to `NORMAL`. This seems to be better in WAL mode.